src/feature/hs/hs_circuit.c

```diff
@@ -1224,3 +1239,41 @@ hs_circ_cleanup(circuit_t *circ)
     hs_circuitmap_remove_circuit(circ);
   }
 }
+
+/** We are about to free this <b>circ</b>. Clean it up from any related HS
+ * data structures. This function can be called multiple times safely for the
+ * same circuit. */
+void
+hs_circ_cleanup_on_free(circuit_t *circ)
+{
+  tor_assert(circ);
+
+  /* NOTE: Bulk of the work of cleaning up a circuit is done here. */
+
+  if (circuit_purpose_is_hs_client(circ->purpose)) {
+    cleanup_on_free_client_circ(circ);
+  }
+
+  /* We have no assurance that the given HS circuit has been closed before and
+   * thus removed from the HS map. This actually happens in unit tests. */
+  if (circ->hs_token) {
```

Comment:
We could in theory call `hs_circuitmap_remove_circuit()` without checking for `hs_token` everytime, since the function does it for us. If we want....

Reply:
New commit for this optimization: 031e35b128

Reply:
I have reverted in ed9d1a240fa728fe because most tests closing/freeing circuits don't call `hs_init()` and thus the HS circuit map asserts in `hs_circuitmap_remove_circuit()`.